# Thesis Github: Gavin Farrell
## Updated: 17/8/21 
## Status; being updated and cleaned 
-> Contact g.farrell13@nuigalway.ie for any specific files or details before then


## Title: Is the a Loss of BGC Diveristy in Colorectal Microbiome
### Description
Given the known dysbiosis of the colorectal gut microbiome in colorectal cancer patients when compared to normal controls, this thesis aimed to determine if there was a further correlation to the number and types of biosynthetic gene clusters (BGC) present in the microbes of the gut microbiome between colorectal cancer patients and normal controls. Data used was short read 100 bp paired end reads. Data was assembled into contigs with Megahit and MetaSpades, and searched for BGC using AntiSMASH and Biosynthetic MetaSpades. Further analsysis and data interpreatation was done using Bash, Python and R scripts.


## 0. Additional files
### 0.1. Sample Accessions and Status
Avaiable on ENA at: https://www.ebi.ac.uk/ena/browser/view/PRJEB12449?show=reads
Accession: PRJEB12449
<br />

### 0.2. Dockerhub container used at: https://hub.docker.com/repository/docker/gfarrell/allin
#### Requires updating as mix of conda environment and singularity (docker base) container used for pipelines
<br />

### 0.3. List of tools used (unavailable)
<br />

## 1. Primary Processing Pipelines:
### Snakemake: Test Pipeline (available)
<br />

### Nextflow: Megahit Pipeline (available)
### Nextflow: MetaSpades Pipeline (available)
<br />

## 2. Bash - File Management (available)
<br />

## 3. Python - File Parsing (available)
<br />

## 4. R - Graphing Data (available)






